File: tests/algorithms/x/Java/other/davis_putnam_logemann_loveland.java
public class Main {
    static class Clause {
        java.util.Map<String,java.math.BigInteger> literals;
        String[] names;
        Clause(java.util.Map<String,java.math.BigInteger> literals, String[] names) {
            this.literals = literals;
            this.names = names;
        }
        Clause() {}
        @Override public String toString() {
            return String.format("{'literals': %s, 'names': %s}", String.valueOf(literals), String.valueOf(names));
        }
    }

    static class EvalResult {
        java.math.BigInteger value;
        Clause clause;
        EvalResult(java.math.BigInteger value, Clause clause) {
            this.value = value;
            this.clause = clause;
        }
        EvalResult() {}
        @Override public String toString() {
            return String.format("{'value': %s, 'clause': %s}", String.valueOf(value), String.valueOf(clause));
        }
    }

    static class Formula {
        Clause[] clauses;
        Formula(Clause[] clauses) {
            this.clauses = clauses;
        }
        Formula() {}
        @Override public String toString() {
            return String.format("{'clauses': %s}", String.valueOf(clauses));
        }
    }

    static class DPLLResult {
        boolean sat;
        java.util.Map<String,java.math.BigInteger> model;
        DPLLResult(boolean sat, java.util.Map<String,java.math.BigInteger> model) {
            this.sat = sat;
            this.model = model;
        }
        DPLLResult() {}
        @Override public String toString() {
            return String.format("{'sat': %s, 'model': %s}", String.valueOf(sat), String.valueOf(model));
        }
    }

    static Clause clause1;
    static Clause clause2;
    static Formula formula;
    static String formula_str;
    static Clause[] clauses;
    static String[] symbols = ((String[])(new String[]{"A4", "A3", "A5", "A1"}));
    static java.util.Map<String,java.math.BigInteger> model = null;
    static DPLLResult result;

    static Clause new_clause(String[] lits) {
        java.util.Map<String,java.math.BigInteger> m = ((java.util.Map<String,java.math.BigInteger>)(new java.util.LinkedHashMap<String, java.math.BigInteger>()));
        String[] names_1 = ((String[])(new String[]{}));
        java.math.BigInteger i_1 = java.math.BigInteger.valueOf(0);
        while (i_1.compareTo(new java.math.BigInteger(String.valueOf(lits.length))) < 0) {
            String lit_1 = lits[_idx((lits).length, ((java.math.BigInteger)(i_1)).longValue())];
m.put(lit_1, (java.math.BigInteger.valueOf(1)).negate());
            names_1 = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(names_1), java.util.stream.Stream.of(lit_1)).toArray(String[]::new)));
            i_1 = i_1.add(java.math.BigInteger.valueOf(1));
        }
        return new Clause(m, ((String[])(names_1)));
    }

    static Clause assign_clause(Clause c, java.util.Map<String,java.math.BigInteger> model) {
        java.util.Map<String,java.math.BigInteger> lits = c.literals;
        java.math.BigInteger i_3 = java.math.BigInteger.valueOf(0);
        while (i_3.compareTo(new java.math.BigInteger(String.valueOf(c.names.length))) < 0) {
            String lit_3 = c.names[_idx((c.names).length, ((java.math.BigInteger)(i_3)).longValue())];
            String symbol_1 = _substr(lit_3, (int)(0L), (int)(2L));
            if (model.containsKey(symbol_1)) {
                java.math.BigInteger value_1 = ((java.math.BigInteger)(model).get(symbol_1));
                if ((_substr(lit_3, (int)(((java.math.BigInteger)(new java.math.BigInteger(String.valueOf(_runeLen(lit_3))).subtract(java.math.BigInteger.valueOf(1)))).longValue()), (int)((long)(_runeLen(lit_3)))).equals("'")) && value_1.compareTo((java.math.BigInteger.valueOf(1)).negate()) != 0) {
                    value_1 = java.math.BigInteger.valueOf(1).subtract(value_1);
                }
lits.put(lit_3, value_1);
            }
            i_3 = i_3.add(java.math.BigInteger.valueOf(1));
        }
c.literals = lits;
        return c;
    }

    static EvalResult evaluate_clause(Clause c, java.util.Map<String,java.math.BigInteger> model) {
        java.math.BigInteger i_4 = java.math.BigInteger.valueOf(0);
        while (i_4.compareTo(new java.math.BigInteger(String.valueOf(c.names.length))) < 0) {
            String lit_5 = c.names[_idx((c.names).length, ((java.math.BigInteger)(i_4)).longValue())];
            String sym_1 = String.valueOf((_substr(lit_5, (int)(((java.math.BigInteger)(new java.math.BigInteger(String.valueOf(_runeLen(lit_5))).subtract(java.math.BigInteger.valueOf(1)))).longValue()), (int)((long)(_runeLen(lit_5)))).equals("'")) ? _substr(lit_5, (int)(0L), (int)(2L)) : lit_5 + "'");
            if (c.literals.containsKey(sym_1)) {
                return new EvalResult(java.math.BigInteger.valueOf(1), c);
            }
            i_4 = i_4.add(java.math.BigInteger.valueOf(1));
        }
        c = assign_clause(c, model);
        i_4 = java.math.BigInteger.valueOf(0);
        while (i_4.compareTo(new java.math.BigInteger(String.valueOf(c.names.length))) < 0) {
            String lit_7 = c.names[_idx((c.names).length, ((java.math.BigInteger)(i_4)).longValue())];
            java.math.BigInteger value_3 = ((java.math.BigInteger)(c.literals).get(lit_7));
            if (value_3.compareTo(java.math.BigInteger.valueOf(1)) == 0) {
                return new EvalResult(java.math.BigInteger.valueOf(1), c);
            }
            if (value_3.compareTo((java.math.BigInteger.valueOf(1)).negate()) == 0) {
                return new EvalResult((java.math.BigInteger.valueOf(1)).negate(), c);
            }
            i_4 = i_4.add(java.math.BigInteger.valueOf(1));
        }
        java.math.BigInteger any_true_1 = java.math.BigInteger.valueOf(0);
        i_4 = java.math.BigInteger.valueOf(0);
        while (i_4.compareTo(new java.math.BigInteger(String.valueOf(c.names.length))) < 0) {
            String lit_9 = c.names[_idx((c.names).length, ((java.math.BigInteger)(i_4)).longValue())];
            if (((java.math.BigInteger)(c.literals).get(lit_9)).compareTo(java.math.BigInteger.valueOf(1)) == 0) {
                any_true_1 = java.math.BigInteger.valueOf(1);
            }
            i_4 = i_4.add(java.math.BigInteger.valueOf(1));
        }
        return new EvalResult(any_true_1, c);
    }

    static Formula new_formula(Clause[] cs) {
        return new Formula(((Clause[])(cs)));
    }

    static String[] remove_symbol(String[] symbols, String s) {
        String[] res = ((String[])(new String[]{}));
        java.math.BigInteger i_6 = java.math.BigInteger.valueOf(0);
        while (i_6.compareTo(new java.math.BigInteger(String.valueOf(symbols.length))) < 0) {
            if (!(symbols[_idx((symbols).length, ((java.math.BigInteger)(i_6)).longValue())].equals(s))) {
                res = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(res), java.util.stream.Stream.of(symbols[_idx((symbols).length, ((java.math.BigInteger)(i_6)).longValue())])).toArray(String[]::new)));
            }
            i_6 = i_6.add(java.math.BigInteger.valueOf(1));
        }
        return ((String[])(res));
    }

    static DPLLResult dpll_algorithm(Clause[] clauses, String[] symbols, java.util.Map<String,java.math.BigInteger> model) {
        boolean all_true = true;
        java.math.BigInteger i_8 = java.math.BigInteger.valueOf(0);
        while (i_8.compareTo(new java.math.BigInteger(String.valueOf(clauses.length))) < 0) {
            EvalResult ev_1 = evaluate_clause(clauses[_idx((clauses).length, ((java.math.BigInteger)(i_8)).longValue())], model);
clauses[(int)(((java.math.BigInteger)(i_8)).longValue())] = ev_1.clause;
            if (ev_1.value.compareTo(java.math.BigInteger.valueOf(0)) == 0) {
                return new DPLLResult(false, ((java.util.Map<String,java.math.BigInteger>)(new java.util.LinkedHashMap<String, java.math.BigInteger>())));
            } else             if (ev_1.value.compareTo((java.math.BigInteger.valueOf(1)).negate()) == 0) {
                all_true = false;
            }
            i_8 = i_8.add(java.math.BigInteger.valueOf(1));
        }
        if (all_true) {
            return new DPLLResult(true, model);
        }
        String p_1 = symbols[_idx((symbols).length, 0L)];
        String[] rest_1 = ((String[])(remove_symbol(((String[])(symbols)), p_1)));
        java.util.Map<String,java.math.BigInteger> tmp1_1 = model;
        java.util.Map<String,java.math.BigInteger> tmp2_1 = model;
tmp1_1.put(p_1, java.math.BigInteger.valueOf(1));
tmp2_1.put(p_1, java.math.BigInteger.valueOf(0));
        DPLLResult res1_1 = dpll_algorithm(((Clause[])(clauses)), ((String[])(rest_1)), tmp1_1);
        if (res1_1.sat) {
            return res1_1;
        }
        return dpll_algorithm(((Clause[])(clauses)), ((String[])(rest_1)), tmp2_1);
    }

    static String str_clause(Clause c) {
        String line = "{";
        boolean first_1 = true;
        java.math.BigInteger i_10 = java.math.BigInteger.valueOf(0);
        while (i_10.compareTo(new java.math.BigInteger(String.valueOf(c.names.length))) < 0) {
            String lit_11 = c.names[_idx((c.names).length, ((java.math.BigInteger)(i_10)).longValue())];
            if (first_1) {
                first_1 = false;
            } else {
                line = line + " , ";
            }
            line = line + lit_11;
            i_10 = i_10.add(java.math.BigInteger.valueOf(1));
        }
        line = line + "}";
        return line;
    }

    static String str_formula(Formula f) {
        String line_1 = "{";
        java.math.BigInteger i_12 = java.math.BigInteger.valueOf(0);
        while (i_12.compareTo(new java.math.BigInteger(String.valueOf(f.clauses.length))) < 0) {
            line_1 = line_1 + String.valueOf(str_clause(f.clauses[_idx((f.clauses).length, ((java.math.BigInteger)(i_12)).longValue())]));
            if (i_12.compareTo(new java.math.BigInteger(String.valueOf(f.clauses.length)).subtract(java.math.BigInteger.valueOf(1))) < 0) {
                line_1 = line_1 + " , ";
            }
            i_12 = i_12.add(java.math.BigInteger.valueOf(1));
        }
        line_1 = line_1 + "}";
        return line_1;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            clause1 = new_clause(((String[])(new String[]{"A4", "A3", "A5'", "A1", "A3'"})));
            clause2 = new_clause(((String[])(new String[]{"A4"})));
            formula = new_formula(((Clause[])(new Clause[]{clause1, clause2})));
            formula_str = String.valueOf(str_formula(formula));
            clauses = ((Clause[])(new Clause[]{clause1, clause2}));
            model = ((java.util.Map<String,java.math.BigInteger>)(new java.util.LinkedHashMap<String, java.math.BigInteger>()));
            result = dpll_algorithm(((Clause[])(clauses)), ((String[])(symbols)), model);
            if (result.sat) {
                System.out.println("The formula " + formula_str + " is satisfiable.");
            } else {
                System.out.println("The formula " + formula_str + " is not satisfiable.");
            }
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{\"duration_us\": " + _benchDuration + ", \"memory_bytes\": " + _benchMemory + ", \"name\": \"main\"}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _substr(String s, int i, int j) {
        int len = _runeLen(s);
        if (i < 0) i = 0;
        if (j > len) j = len;
        if (i > j) i = j;
        int start = s.offsetByCodePoints(0, i);
        int end = s.offsetByCodePoints(0, j);
        return s.substring(start, end);
    }

    static int _idx(int len, long i) {
        return (int)(i < 0 ? len + i : i);
    }
}

File: tests/algorithms/transpiler/FS/other/davis_putnam_logemann_loveland.fs
// Generated 2025-08-22 13:05 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
type Clause = {
    mutable _literals: System.Collections.Generic.IDictionary<string, int>
    mutable _names: string array
}
type EvalResult = {
    mutable _value: int
    mutable _clause: Clause
}
type Formula = {
    mutable _clauses: Clause array
}
type DPLLResult = {
    mutable _sat: bool
    mutable _model: System.Collections.Generic.IDictionary<string, int>
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System.Collections.Generic

let rec new_clause (lits: string array) =
    let mutable __ret : Clause = Unchecked.defaultof<Clause>
    let mutable lits = lits
    try
        let mutable m: System.Collections.Generic.IDictionary<string, int> = _dictCreate []
        let mutable _names: string array = Array.empty<string>
        let mutable i: int = 0
        while i < (Seq.length (lits)) do
            let lit: string = _idx lits (int i)
            m <- _dictAdd (m) (string (lit)) (0 - 1)
            _names <- Array.append _names [|lit|]
            i <- i + 1
        __ret <- { _literals = m; _names = _names }
        raise Return
        __ret
    with
        | Return -> __ret
and assign_clause (c: Clause) (_model: System.Collections.Generic.IDictionary<string, int>) =
    let mutable __ret : Clause = Unchecked.defaultof<Clause>
    let mutable c = c
    let mutable _model = _model
    try
        let mutable lits: System.Collections.Generic.IDictionary<string, int> = c._literals
        let mutable i: int = 0
        while i < (Seq.length (c._names)) do
            let lit: string = _idx (c._names) (int i)
            let symbol: string = _substring lit 0 2
            if _model.ContainsKey(symbol) then
                let mutable _value: int = _dictGet _model ((string (symbol)))
                if ((_substring lit ((String.length (lit)) - 1) (String.length (lit))) = "'") && (_value <> (0 - 1)) then
                    _value <- 1 - _value
                lits <- _dictAdd (lits) (string (lit)) (_value)
            i <- i + 1
        c._literals <- lits
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
and evaluate_clause (c: Clause) (_model: System.Collections.Generic.IDictionary<string, int>) =
    let mutable __ret : EvalResult = Unchecked.defaultof<EvalResult>
    let mutable c = c
    let mutable _model = _model
    try
        let mutable i: int = 0
        while i < (Seq.length (c._names)) do
            let lit: string = _idx (c._names) (int i)
            let sym: string = if (_substring lit ((String.length (lit)) - 1) (String.length (lit))) = "'" then (_substring lit 0 2) else (lit + "'")
            if (c._literals).ContainsKey(sym) then
                __ret <- { _value = 1; _clause = c }
                raise Return
            i <- i + 1
        c <- assign_clause (c) (_model)
        i <- 0
        while i < (Seq.length (c._names)) do
            let lit: string = _idx (c._names) (int i)
            let mutable _value: int = _dictGet (c._literals) ((string (lit)))
            if _value = 1 then
                __ret <- { _value = 1; _clause = c }
                raise Return
            if _value = (0 - 1) then
                __ret <- { _value = 0 - 1; _clause = c }
                raise Return
            i <- i + 1
        let mutable any_true: int = 0
        i <- 0
        while i < (Seq.length (c._names)) do
            let lit: string = _idx (c._names) (int i)
            if (_dictGet (c._literals) ((string (lit)))) = 1 then
                any_true <- 1
            i <- i + 1
        __ret <- { _value = any_true; _clause = c }
        raise Return
        __ret
    with
        | Return -> __ret
and new_formula (cs: Clause array) =
    let mutable __ret : Formula = Unchecked.defaultof<Formula>
    let mutable cs = cs
    try
        __ret <- { _clauses = cs }
        raise Return
        __ret
    with
        | Return -> __ret
and remove_symbol (symbols: string array) (s: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable symbols = symbols
    let mutable s = s
    try
        let mutable res: string array = Array.empty<string>
        let mutable i: int = 0
        while i < (Seq.length (symbols)) do
            if (_idx symbols (int i)) <> s then
                res <- Array.append res [|(_idx symbols (int i))|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and dpll_algorithm (_clauses: Clause array) (symbols: string array) (_model: System.Collections.Generic.IDictionary<string, int>) =
    let mutable __ret : DPLLResult = Unchecked.defaultof<DPLLResult>
    let mutable _clauses = _clauses
    let mutable symbols = symbols
    let mutable _model = _model
    try
        let mutable all_true: bool = true
        let mutable i: int = 0
        while i < (Seq.length (_clauses)) do
            let ev: EvalResult = evaluate_clause (_idx _clauses (int i)) (_model)
            _clauses <- _arrset _clauses (int i) (ev._clause)
            if (ev._value) = 0 then
                __ret <- { _sat = false; _model = _dictCreate<string, int> [] }
                raise Return
            else
                if (ev._value) = (0 - 1) then
                    all_true <- false
            i <- i + 1
        if all_true then
            __ret <- { _sat = true; _model = _model }
            raise Return
        let p: string = _idx symbols (int 0)
        let rest: string array = remove_symbol (symbols) (p)
        let mutable tmp1: System.Collections.Generic.IDictionary<string, int> = _model
        let mutable tmp2: System.Collections.Generic.IDictionary<string, int> = _model
        tmp1 <- _dictAdd (tmp1) (string (p)) (1)
        tmp2 <- _dictAdd (tmp2) (string (p)) (0)
        let res1: DPLLResult = dpll_algorithm (_clauses) (rest) (tmp1)
        if res1._sat then
            __ret <- res1
            raise Return
        __ret <- dpll_algorithm (_clauses) (rest) (tmp2)
        raise Return
        __ret
    with
        | Return -> __ret
and str_clause (c: Clause) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable c = c
    try
        let mutable line: string = "{"
        let mutable first: bool = true
        let mutable i: int = 0
        while i < (Seq.length (c._names)) do
            let lit: string = _idx (c._names) (int i)
            if first then
                first <- false
            else
                line <- line + " , "
            line <- line + lit
            i <- i + 1
        line <- line + "}"
        __ret <- line
        raise Return
        __ret
    with
        | Return -> __ret
and str_formula (f: Formula) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable f = f
    try
        let mutable line: string = "{"
        let mutable i: int = 0
        while i < (Seq.length (f._clauses)) do
            line <- line + (str_clause (_idx (f._clauses) (int i)))
            if i < ((Seq.length (f._clauses)) - 1) then
                line <- line + " , "
            i <- i + 1
        line <- line + "}"
        __ret <- line
        raise Return
        __ret
    with
        | Return -> __ret
let clause1: Clause = new_clause (unbox<string array> [|"A4"; "A3"; "A5'"; "A1"; "A3'"|])
let clause2: Clause = new_clause (unbox<string array> [|"A4"|])
let formula: Formula = new_formula (unbox<Clause array> [|clause1; clause2|])
let formula_str: string = str_formula (formula)
let _clauses: Clause array = unbox<Clause array> [|clause1; clause2|]
let symbols: string array = unbox<string array> [|"A4"; "A3"; "A5"; "A1"|]
let mutable _model: System.Collections.Generic.IDictionary<string, int> = _dictCreate []
let result: DPLLResult = dpll_algorithm (_clauses) (symbols) (_model)
if result._sat then
    ignore (printfn "%s" (("The formula " + formula_str) + " is satisfiable."))
else
    ignore (printfn "%s" (("The formula " + formula_str) + " is not satisfiable."))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
